test/common/network/udp_listener_impl_test.cc

```diff
@@ -352,68 +361,98 @@ TEST_P(UdpListenerImplTest, SendData) {
   const std::string payload("hello world");
   Buffer::InstancePtr buffer(new Buffer::OwnedImpl());
   buffer->add(payload);
-  UdpSendData send_data{client_socket_->localAddress(), *buffer};
-
-  auto send_result = listener_->send(send_data);
+  // Use a self address that is unlikely to be picked by source address discovery
+  // algorithm if not specified in recvmsg. Port is not taken into
+  // consideration.
+  Address::InstanceConstSharedPtr send_from_addr;
+  if (version_ == Address::IpVersion::v4) {
+    // Linux kernel regards any 127.x.x.x as local address. But Mac OS doesn't.
+    send_from_addr.reset(new Address::Ipv4Instance(
+#ifndef __APPLE__
+        "127.1.2.3",
+#else
+        "127.0.0.1",
+#endif
+        server_socket_->localAddress()->ip()->port()));
+  } else {
+    // Only use non-local v6 address if IP_FREEBIND is supported. Otherwise use
+    // ::1 to avoid EINVAL error. Unfortunately this can't verify that sendmsg with
+    // customized source address is doing the work because kernel also picks ::1
+    // if it's not specified in cmsghdr.
+    send_from_addr.reset(new Address::Ipv6Instance(
+#ifdef IP_FREEBIND
+        "::9",
```

Comment:
@danzh2010 this one cause problems during Google Envoy import. It seems we don't support the ability to bind to ::9 like this on our internal test farm. One of the issues that seems to exist here is that we assume that a compile time support for IP_FREEBIND translates to runtime ability. Do you think you could express this differently? CC @dnoe @yanavlasov

Reply:
> @danzh2010 this one cause problems during Google Envoy import. It seems we don't support the ability > to bind to ::9 like this on our internal test farm. One of the issues that seems to exist here is that we
> assume that a compile time support for IP_FREEBIND translates to runtime ability. 

This might has walk-around, I'll check our test infrastructure about it.

> Do you think you could express this differently? CC @dnoe @yanavlasov

It's hard to test it in a different way as we need an v6 address that we can use to send which is different from what kernel would have picked for us. An compromise is we always use ::1 for v6, but this will leave v6 behavior untested.